### ISSUE TITLE
Allow can-observe to make plain Objects and Arrays observable and work without proxy

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/can-observe#19]
__tldr;__ Instead of creating a DefineMap or DefineList, allow people to make an Object or Array observable. Benefits:

- These observables will work with lodash and other POJO helpers.
- They ARE Object and Arrays.
- We reference the same object.

> This was discussed at a [contributors meeting (11:35)](https://youtu.be/95OX6-qii24?t=11m35s)

## The Problem

1. Creating `DefineMap` and `DefineList` is a lot of cognitive overhead for a new developer.  Ideally, new users won't need to know CanJS observables existed and instead, we will just make everything they gave us an observable.

2. The current proxy method isn't supported in all browsers and messes up references.  For example:
   ```js
   var observe = require("can-observe");
   var obj = {};
   var observedObj = observe(obj);
   obj === observedObj //-> false
   ```
   This can create all sorts of difficult to understand issues.

## The Solution

I propose making `can-observe`, when called with `true` as the second argument, mutate that argument to make it observable.  For example:

```js
var observe = require("can-observe");
var obj = {foo: "bar"};

var observedObj = observe(obj, true);
obj === observedObj //-> true

canReflect.onKeyValue(obj, "foo", function(newValue, oldValue){
  newValue //-> "zed"
})

obj.foo = "zed" 
```

This will work similar with arrays:

```js
var observe = require("can-observe");
var hobbies = ["first","second","third"];

observe(hobbies, true);

canReflect.onKeyValue(hobbies, "length", function(newValue, oldValue){
  newValue //-> 4
})

hobbies.push("super staring");
```

Finally, if `can-observe` is called with `true` as the second argument and `Proxy` exists, it will fall back to the current behavior (that supports dynamic properties):

```js
var observe = require("can-observe");
var obj = {};
var observedObj = observe(obj, true);
observedObj.foo // calls Observation.add(observedObj, "foo");
```

### Limitations

Note that adding new properties to these objects after they have been passed to `observe` will not be observable by computes.  For example:

```js
var observe = require("can-observe");
var obj = observe({first: "Justin"});

var fullName = compute(function(){
  return obj.first + " " + obj.last
});

fullName.on("change", function(){
  // not called!
})

obj.last = "Meyer"
```

We have two options to deal with this situation:

- `Object.seal()` these objects so users get warnings when adding new properties that were not already defined.  Users will need to define their properties, or use the proxy version. Or,
- Tell users to use `canReflect.get(obj, "last")` if the property might not be already defined.

## Technical Details

To make this work for objects we will:

1.  Attach a symbol to the object where all event handlers and other observable related data will go.
2.  Loop through every property and add a `getter/setter` similar to those of "any" type in `can-define`.
3.  Need to make Observation.add work without `_cid`.

To make this work for Arrays we will:

1.  Create a prototype with all mutate-able methods on an array similar to DefineList.
2.  Assign the Array's __proto__ to this prototype.

**Comment by justinbmeyer** [comment]
I confirmed that this does not work in VueJS, which uses a similar technique:

```
var app = new Vue({
  el: '#app',
  data: {
    foo: {}
  },
  methods: {
    reverseMessage: function () {
      this.foo.bar ="works"
    }
  }
})
```

This likely proves that the problem of undefined properties is unavoidable w/o proxy.

**Comment by justinbmeyer** [comment]
- only change POJOs.

```
var obj  = { map : new DefineMap() }
```

```js
mutableArrayPrototype = Object.create(Array.prototype);
mutableArrayPrototype.push = function(){
  // fire `onValue` handlers 
  Array.prototype.push.apply(this, arguments);
}
mutatableArrayPrototype.forEach = function(){
  Observation.add(this);
  Array.prototype.forEach.apply(this, arguments);
}

observeArray = function(array){
    Object.setPrototypeOf(array, mutableArrayPrototype);

    // XXXXX make an observable-like property XXXXX
    // Object.defineProperty(array,"length", {});

    // might want to call `observeObject()` only on non-index properties

    // go through items, check if they are arrays or objects and PUNCH them (recurse)
}



var observableDataSymbol = canSymbol.for("can.observeData")
observeObject = function(obj) {
    if(obj[observableDataSymbol]) {
        return obj;
    } else {
        Object.defineProperty(obj, "_cid", {
            value: cid({})._cid,
            enumerable: false
        })
        // on both enumerable and non-enumerable? - check if vue
        for(var prop in obj) {
            // what if a getter / setter already there
            var data = obj[observableDataSymbol] = {};
            var value = obj[prop];
            data[prop] = ifObject ? observeObject( value ) : value;
            Object.defineProperty(obj,prop, {
                get: function(){
            		Observation.add(target, name);
                        value = data[name]
                        if(Array.isArray( value) ) Observation.add(value)
        		return value;
                },
                set: function(value){
                    var old = data[key];
                    // TODO: if value is an object, or array, PUNCH -> observeObject(obj)
        			var change = old !== value;
        			if(change) {
        				data[key] = value;

                        canBatch.queue(YOUR_HANDLERS)
        				/**canBatch.dispatch.call(this, {
        					type: key,
        					target: target
        				}, [value, old]);*/

            			return true;
                    }
                }
            })
        }
        canReflect.assignSymbols({
            "can.onKeyValue": function(key, handler){
                ON_SOMETHING.push(handler)
            },
            "can.offKeyValue": function(){
                
            }
        })
    }
}
```

**Comment by justinbmeyer** [comment]
```js
var arr = observeArray([]);

var defineMap =  new DefineMap({
  myArray: arr
});

var obj = observeObj({myArray: arr})

defineMap.myArray //-> DefineList

var length2 = c(function(){
  var count = 0;
  defineMap.myArray.forEach(function(){
    count++;
  });
  count++;
})

var length = c(function(){
  // Observation.add(defineMap, "myArray");
  return defineMap.myArray.length
});

var length3 = c(function(){
  
  return obj.myArray.length
});

defineMap.myArray.pop();
```
- [ ] - DefineMap & DefineList will [] convert to DefineList and {} to DefineMap.  Fix `canReflect.isObservable(arr)` -> true && `canReflect.isPlainObject(arr)` -> false?
- [ ] - Other types will not call `Observation.add(array)` when the array is being referenced.

**Comment by justinbmeyer** [comment]
```js
DefineMap

getter(prop) {
  
  Observation.add(this, prop);
  value = data[name]
  // Option 1: use isValueLike
  if( canReflect.isValueLike(value) ) {
    Observation.add(value) 
    // this doesn't make sense for computes because when read they always call Observation.add themselves
  } 
  // Option 2: use a new canReflect value
  if( canReflect.isValueLikeButReadableWithoutObservationAdd(value)  ) {
    Observation.add(value)
  }
  // Option 3: like #2, but make can-observation do it:
  Observation.add(this, prop, value);
  return value;
}
```
